File: backend/app/schemas/rules.py
from typing import List, Optional
from pydantic import BaseModel, Field, ConfigDict

class RuleModel(BaseModel):
    rule_id: str = Field(..., description="Unique legal rule identifier (e.g. LM-MRP-001)")
    name: str = Field(..., description="Human-readable rule name")
    description: str = Field(..., description="Summary of the rule objective")
    requirement: str = Field(..., description="Precise statutory requirement for compliance")
    applicable_product_categories: List[str] = Field(
        default_factory=lambda: ["all"],
        description="Target categories (e.g. ['all'], ['food'], ['beverages'])"
    )
    field_to_validate: str = Field(..., description="Target declaration field (e.g. mrp, net_quantity)")
    validation_type: str = Field(..., description="Type of check: REQUIRED, FORMAT, UNIT_SPECIFICATION, etc.")
    severity: str = Field(..., description="Violation severity level: CRITICAL, HIGH, MEDIUM, LOW")
    effective_from: str = Field(..., description="Effective start date (YYYY-MM-DD)")
    effective_until: Optional[str] = Field(None, description="Sunset or superseded date (YYYY-MM-DD), or null")
    rule_version: str = Field(..., description="Version of the codified rule engine catalog (e.g. 2026.1)")
    source_reference: str = Field(..., description="Official citation under Legal Metrology Act / Rules")
    enabled: bool = Field(True, description="Whether this rule is currently evaluated")

    model_config = ConfigDict(from_attributes=True)

class RuleListResponse(BaseModel):
    rules: List[RuleModel] = Field(..., description="List of matching rule definitions")
    total: int = Field(..., description="Total count of returned rules")
    selected_version: str = Field(..., description="Active or queried rule catalog version")
    available_versions: List[str] = Field(..., description="All discoverable rule catalog versions")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/rules.ts
export interface RuleModel {
  rule_id: string;
  name?: string;
  title?: string;
  description?: string;
  requirement: string;
  applicable_product_categories?: string[];
  applicable_categories?: string[];
  field_to_validate: string;
  validation_type: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  effective_from: string;
  effective_until: string | null;
  rule_version: string;
  section?: string;
  sub_rule?: string;
  source_reference?: string;
  source_url?: string;
  source_page?: string;
  status?: string;
  enabled?: boolean;
}

export interface RuleListResponse {
  rules: RuleModel[];
  total: number;
  selected_version: string;
  available_versions: string[];
}

export interface RegulatoryDocumentDTO {
  id: string;
  document_name: string;
  document_type: string;
  notification_number?: string;
  publication_date: string;
  effective_date: string;
  source_url?: string;
  source_reference: string;
  content_hash?: string;
  version: string;
  status: string;
}

export interface RegulatoryRuleDTO {
  id: string;
  rule_id: string;
  rule_version: string;
  title: string;
  section: string;
  sub_rule?: string;
  requirement: string;
  applicable_categories: string[];
  field_to_validate: string;
  validation_type: string;
  validation_expression?: Record<string, any>;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  effective_from: string;
  effective_until?: string | null;
  source_document_id?: string;
  source_url?: string;
  source_page?: string;
  source_excerpt?: string;
  status: 'PENDING_REVIEW' | 'APPROVED' | 'ACTIVE' | 'SUPERSEDED' | 'REJECTED';
}

export interface RuleAmendmentDTO {
  id: string;
  document_id: string;
  rule_id: string;
  change_type: string;
  previous_value?: string | null;
  new_value: string;
  effective_from: string;
  effective_until?: string | null;
  explanation?: string | null;
  created_at?: string;
}

export interface RegulatoryCatalogSummary {
  total_rules: number;
  active_rules: number;
  pending_rules: number;
  superseded_rules: number;
  documents_count: number;
  amendments_count: number;
  available_versions: string[];
  latest_version: string;
}
